srml/contracts/src/lib.rs

```diff
@@ -702,34 +703,41 @@ decl_module! {
 			} else {
 				origin_contract.last_write
 			};
-
-			let key_values_taken = delta.iter()
-				.filter_map(|key| {
-					child::get_raw(&origin_contract.trie_id, &blake2_256(key)).map(|value| {
-						child::kill(&origin_contract.trie_id, &blake2_256(key));
-						(key, value)
+			if let Some(child_trie) = child::child_trie(
+				&prefixed_trie_id(&origin_contract.trie_id[..])[..]
+			) {
+
+				let key_values_taken = delta.iter()
+					.filter_map(|key| {
+						child::get_raw(child_trie.node_ref(), &blake2_256(key)).map(|value| {
+							child::kill(&child_trie, &blake2_256(key));
+							(key, value)
+						})
 					})
-				})
-				.collect::<Vec<_>>();
+					.collect::<Vec<_>>();
 
-			let tombstone = <TombstoneContractInfo<T>>::new(
-				// This operation is cheap enough because last_write (delta not included)
-				// is not this block as it has been checked earlier.
-				&runtime_io::child_storage_root(&origin_contract.trie_id)[..],
-				code_hash,
-			);
+				let tombstone = <TombstoneContractInfo<T>>::new(
+					// This operation is cheap enough because last_write (delta not included)
+					// is not this block as it has been checked earlier.
+					&runtime_io::child_storage_root(&child_trie)[..],
+					code_hash,
+				);
+
+				if tombstone != dest_tombstone {
+					for (key, value) in key_values_taken {
+						child::put_raw(&child_trie, &blake2_256(key), &value);
+					}
 
-			if tombstone != dest_tombstone {
-				for (key, value) in key_values_taken {
-					child::put_raw(&origin_contract.trie_id, &blake2_256(key), &value);
+					return Err("Tombstones don't match");
 				}
 
-				return Err("Tombstones don't match");
-			}
+				origin_contract.storage_size -= key_values_taken.iter()
+					.map(|(_, value)| value.len() as u32)
+					.sum::<u32>();
 
-			origin_contract.storage_size -= key_values_taken.iter()
-				.map(|(_, value)| value.len() as u32)
-				.sum::<u32>();
+			} else {
+				return Err("Corrupted contract is missing child storage trie");
```

Comment:
nit: Style-wise, would prefer:

```rust
let child_trie = child::child_trie(&prefixed_trie_id(&origin_contract.trie_id[..])[..])
    .ok_or_else(|| "Corrupted contract is missing child storage trie")?;
```

Which makes it a bit more immediately clear that the child trie is expected to exist and removes the indentation.